x-pack/plugins/discover_log_explorer/common/runtime_types.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as rt from 'io-ts';
+import { fold } from 'fp-ts/lib/Either';
+import { identity } from 'fp-ts/lib/function';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { Context, Errors, IntersectionType, Type, UnionType, ValidationError } from 'io-ts';
+import { isEmpty, omitBy } from 'lodash';
+
+type ErrorFactory = (message: string) => Error;
+
+const getErrorPath = ([first, ...rest]: Context): string[] => {
+  if (typeof first === 'undefined') {
+    return [];
+  } else if (first.type instanceof IntersectionType) {
+    const [, ...next] = rest;
+    return getErrorPath(next);
+  } else if (first.type instanceof UnionType) {
+    const [, ...next] = rest;
+    return [first.key, ...getErrorPath(next)];
+  }
+
+  return [first.key, ...getErrorPath(rest)];
+};
+
+const getErrorType = ({ context }: ValidationError) =>
+  context[context.length - 1]?.type?.name ?? 'unknown';
+
+const formatError = (error: ValidationError) =>
+  error.message ??
+  `in ${getErrorPath(error.context).join('/')}: ${JSON.stringify(
+    error.value
+  )} does not match expected type ${getErrorType(error)}`;
+
+export const formatErrors = (errors: ValidationError[]) =>
+  `Failed to validate: \n${errors.map((error) => `  ${formatError(error)}`).join('\n')}`;
+
+export const createPlainError = (message: string) => new Error(message);
+
+export const throwErrors = (createError: ErrorFactory) => (errors: Errors) => {
+  throw createError(formatErrors(errors));
+};
+
+export const decodeOrThrow =
+  <DecodedValue, EncodedValue, InputValue>(
+    runtimeType: Type<DecodedValue, EncodedValue, InputValue>,
+    createError: ErrorFactory = createPlainError
+  ) =>
+  (inputValue: InputValue) =>
+    pipe(runtimeType.decode(inputValue), fold(throwErrors(createError), identity));
+
+export const removeEmptyStringPropsRT = <Codec extends rt.Mixed>(codec: Codec) =>
+  new rt.Type<rt.TypeOf<Codec>, rt.OutputOf<Codec>, rt.UnknownRecordC>(
+    `removeEmptyStringPropsRT<${codec.name}>`,
+    codec.is,
+    (input, context) => {
+      const sanitizedInput = omitBy(input, isEmpty) as rt.TypeOf<Codec>;
+      return codec.validate(sanitizedInput, context);
+    },
+    (input) => omitBy(input, isEmpty) as rt.TypeOf<Codec>
```

Comment:
This short-circuits all `encode()` calls of the wrapped type. We should delegate to its `encode` method. Can you help me understand why we need it at all?

Reply:
I just noticed this and fix it.
I added this cause the search component will send by default the `nameQuery`or `datasetQuery` as empty strings if the user performs a search but doesn't have any written text. I didn't want to send it to the server as an empty query param, so I need to filter it out somewhere if it is an empty string. There is any place where you would recommend removing the param when is an empty string?

Reply:
Ah, I see - I was confused because I only focused on the `searchAfter` property :facepalm:

Would it make sense to define these as `NonEmptyString`s (https://github.com/elastic/kibana/blob/main/packages/kbn-io-ts-utils/src/non_empty_string_rt/index.ts#L16) and perform the filtering where the state machine action calls `datasetsClient.findIntegrations()`? That way there is an explicit point when `IntegrationSearchParams` get turned into `FindIntegrationsRequestQuery`. :thinking: